### ISSUE TITLE
Introduce error category [unsafe-overload].

### ISSUE THREAD

**Issue by randolf-scholz** [python/mypy#16060]
**Feature**

Specifically, consider this [example from the documentation](https://mypy.readthedocs.io/en/stable/more_types.html#type-checking-the-variants) [[mypy-play]](https://mypy-play.net/?mypy=latest&python=3.11&gist=e3ef44b3191567e27dc1b2887c4e722e):

```python
from typing import overload, Union

@overload
def unsafe_func(x: int) -> int: ...  # [misc] overlap with incompatible return types
@overload
def unsafe_func(x: object) -> str: ...
def unsafe_func(x: object) -> Union[int, str]:
    if isinstance(x, int):
        return 42
    else:
        return "some string"
```

**Pitch**

`misc` is too generic and makes it difficult to specifically disable this kind of error. I propose to give it a specific name like `unsafe-overload`.

Due to the absence of [`Intersection`](https://github.com/python/typing/issues/213) and [`Not`](https://github.com/python/typing/issues/801), this check raises false positives in certain code, for example:

```python
@overload
def foo(x: Vector) -> Vector: ...
@overload
def foo(x: Iterable[Vector]) -> list[Vector]: ...
def foo(x):
   if isinstance(x, Vector):
      return x
   return list(map(foo, x))
```

Where `Vector` implements `__iter__`. Obviously, what we really mean here is:

```python
@overload
def foo(x: Vector) -> Vector: ...
@overload
def foo(x: Iterable[Vector] & Not[Vector]) -> list[Vector]: ...
def foo(x):
   if isinstance(x, Vector):
      return x
   return list(map(foo, x))
```

Which is actually implicit by the ordering of overloads. The latter would not raise `unsafe-overload`, since `Iterable[Vector] & Not[Vector]` is not a subtype of `Vector`. But since it can't be expressed in the current type system, there needs to be a simple way to silence this check.

**Comment by JelleZijlstra** [comment]
Seems fine to me, please submit a PR. In general we should minimize use of the "misc" error code.

One concern is about backward incompatibility: users who use `# type: ignore[misc]` will have to change their code. Not sure if we have a stance about that.

**Comment by randolf-scholz** [comment]
> # This is a catch-all for remaining uncategorized errors.
> MISC: Final = ErrorCode("misc", "Miscellaneous other checks", "General")

Wouldn't it make sense then that any `type: ignore` comment should cover `misc` to ensure forward compatibility?

**Comment by randolf-scholz** [comment]
So if `misc` stands for uncategorized errors, then possible `type: ignore[misc]` should cover any other error code, but emit a warning if it covers an error with a code different from `misc`?

**Comment by AlexWaygood** [comment]
> One concern is about backward incompatibility: users who use `# type: ignore[misc]` will have to change their code. Not sure if we have a stance about that.

We now have a feature where we can make new errors codes "subcodes" of other codes, so that people's existing type: ignore comments don't break. Am example is the newish method-assign error code. @randolf-scholz, I recommend that you update your PR to use a similar approach. 

(This change is a great suggestion by the way, thanks for taking it on.)
